### PR TITLE
Add Zephyr driver api

### PR DESCRIPTION
**Pull request by jukkar**
This is very much WIP.

### CODE REVIEW

File: src/drivers/driver.h
Comment (GannicusGuo) [inline src/drivers/driver.h:1]: why change open source wpa_driver_ops structure to add this get_conn_info?

File: wpa_supplicant/bss.c
Comment (GannicusGuo) [inline wpa_supplicant/bss.c:1]: why change open source bss.c add this wpa_bss_check_pending_connect?
will this change merge to https://w1.fi/cgit/hostap/?

File: wpa_supplicant/wpa_supplicant_i.h
Comment (GannicusGuo) [inline wpa_supplicant/wpa_supplicant_i.h:1]: why change open source structure wpa_supplicant to add "connection_a, connection_b, connection_g "

File: wpa_supplicant/bss.c
Comment (krish2718) [reply wpa_supplicant/bss.c:1]: This was submitted [upstream](https://patchwork.ozlabs.org/project/hostap/patch/20230420192821.376828-1-krishna.t@nordicsemi.no/) but not yet merged.

File: wpa_supplicant/wpa_supplicant_i.h
Comment (krish2718) [reply wpa_supplicant/wpa_supplicant_i.h:1]: This is in the list of things to submit upstream but not done yet.

File: src/drivers/driver.h
Comment (krish2718) [reply src/drivers/driver.h:1]: This is in the list of things to submit upstream but not done yet.

File: wpa_supplicant/wpa_cli_cmds.c [hunk 1]
```diff
@@ -0,0 +1,3695 @@
+/*
```
Comment (krish2718) [inline wpa_supplicant/wpa_cli_cmds.c:1]: FYI, This change to split generic commands in wpa_cli.c in to wpa_cli_cmds.c also was planned to be submitted upstream, but haven't gotten around to that.

File: wpa_supplicant/wpa_cli_zephyr.c [hunk 2]
```diff
@@ -37,7 +37,7 @@ extern struct wpa_global *global;
 
 static void wpa_cli_msg_cb(char *msg, size_t len)
 {
-	wpa_printf(MSG_INFO, "%s\n", msg);
```
Comment (krish2718) [inline wpa_supplicant/wpa_cli_zephyr.c]: can this commit be squashed in to the `fix newline` commit?

**Comment by dleach02** [comment]
FYI, if we are doing commits here and submitting changes to the hostap upstream, we really should consider adopting the notation Nordic uses for engaging with upstream projects... e.g. [zephyr fromlist] 

It will make it clear to the reviewers what is going on.

@carlescufi, thoughts?

File: src/drivers/driver.h
Comment (gannicus-gj) [reply src/drivers/driver.h:1]: Hi @krish2718 , is it that hostap (upstream) needs to adopt the "list of things to submit" first,
before merging this "hostap/pull/2" to zephyr?  or there is no such dependency?

**Comment by gannicus-gj** [comment]
Hi @krish2718 , is it that hostap (upstream) needs to adopt the "list of things to submit" first,
before merging this "hostap/pull/2" to zephyr? or there is no such dependency?

**Comment by jukkar** [comment]
> FYI, if we are doing commits here and submitting changes to the hostap upstream, we really should consider adopting the notation Nordic uses for engaging with upstream projects... e.g. [zephyr fromlist]

That is fine by me, we can certainly do that if needed. Although that means that we would need to do rebases in this repo, that is fine too but we do not seem to do that in other zephyr repos.

**Comment by dleach02** [comment]
I'm pointing this out because of the number of comments/responses around changes that have been submitted to the upstream. Most of the other modules we have in Zephyr usually don't have too many changes like this and I'm not sure if hostap will be different in this way.

Again, just a suggestion. I like the Nordic tagging/workflow as it does clearly identify the origin of a commit.

**Comment by krish2718** [comment]
> Again, just a suggestion. I like the Nordic tagging/workflow as it does clearly identify the origin of a commit.

Agreed, this would help in maintaining and doing regular rebases to upstream. But we need to pull in few tools/repos from NCS to make it happen, but I guess we could start with just the `sauce tags`.

**Comment by jukkar** [comment]
We (me, Carles and Robert) pondered the commit tagging internally and found out that:
* Rebasing and rewriting history is not allowed according to this zephyr policy https://docs.zephyrproject.org/latest/develop/modules.html#allowed-practices
* It seems that only `[toup]` and `[noup]` make sense with upstream hostap (those tags could be named also `[zephyr toup]` or `[zephyr noup]`)
* If the commit is accepted to upstream (the `[toup]` tag was used), then when merging the upstream back, we would need to revert the original commit (because the commit message is not the same)

We can certainly do this, if needed. WDYT?

**Comment by jukkar** [comment]
If we want to get the hostap updates here and in the glue PR in zephyr repo merged to 3.6, then we would need to speed up the review as merge window closes next week. @dleach02 please prioritize this.

File: src/utils/os_zephyr.c [hunk 3]
```diff
@@ -191,7 +191,7 @@ int os_fdatasync(FILE *stream)
 
 char *os_strdup(const char *s)
 {
-	return strdup(s);
+	return os_memdup(s, strlen(s) + 1);
```
Comment (krish2718) [inline src/utils/os_zephyr.c:194]: interesting, may be leave a TODO here?

**Review by krish2718 — APPROVED** [review]

**Comment by dleach02** [comment]
> We (me, Carles and Robert) pondered the commit tagging internally and found out that:
> 
> * Rebasing and rewriting history is not allowed according to this zephyr policy [docs.zephyrproject.org/latest/develop/modules.html#allowed-practices](https://docs.zephyrproject.org/latest/develop/modules.html#allowed-practices)
> * It seems that only `[toup]` and `[noup]` make sense with upstream hostap (those tags could be named also `[zephyr toup]` or `[zephyr noup]`)
> * If the commit is accepted to upstream (the `[toup]` tag was used), then when merging the upstream back, we would need to revert the original commit (because the commit message is not the same)
> 
> We can certainly do this, if needed. WDYT?

I'm fine with that. I actually think that is how Nordic typically does this anyway when doing the mergeup; the PR has a bunch of reverts then the pull in of the upstream, followed by reapply of those items that are still pending. But you guys know this process better then I do ;-)

I just feel that this project appears to be going down a path of having lots of changes to the upstream and I think the Nordic workflow will help avoid technical debt in the long run.

**Review by dleach02 — APPROVED** [review]

**Comment by krish2718** [comment]
> The base branch requires all commits to be signed.

I guess this was accidental and should be disabled? I don't think Zephyr enforces `signed commits`.

**Pull request merged**